### PR TITLE
Cache node positions

### PR DESCRIPTION
**Pull request by maxbrunsfeld**
This also adds a `compare` operation, so that markers can be sorted without having to create any JS objects from C++ code.

### CODE REVIEW

File: src/native/marker-index.cc [hunk 1]
```diff
@@ -246,18 +262,26 @@ unordered_map<MarkerId, Range> MarkerIndex::Dump() {
...
+    while (current_node->parent) {
+      if (current_node->parent->right == current_node) {
+        position = current_node->parent->left_extent.Traverse(position);
+      }
 
-    node = node->parent;
+      current_node = current_node->parent;
+    }
+    node_position_cache.insert({node, position});
+    return position;
```
Comment (nathansobo) [inline src/native/marker-index.cc]: Might be nice to add an `else` branch here to have 2 explicit return paths.

File: src/native/marker-index.cc [hunk 2]
```diff
@@ -246,18 +262,26 @@ unordered_map<MarkerId, Range> MarkerIndex::Dump() {
...
   }
-  return position;
+
+  return cache_entry->second;
 }
 
 void MarkerIndex::DeleteNode(Node *node) {
+  node_position_cache.erase(node);
```
Comment (nathansobo) [inline src/native/marker-index.cc]: :heavy_check_mark:

File: test/marker-index.benchmark.js [hunk 3]
```diff
@@ -24,7 +25,13 @@ function runBenchmark () {
     enqueueSplice()
     enqueueDelete()
   }
+
+  for (let i = 0; i < 500; i++) {
+    enqueueRangeQuery()
```
Comment (nathansobo) [inline test/marker-index.benchmark.js]: I should have benchmarked this from the beginning. Lesson learned. Would be interested in testing different queries even but this is a good start.

**Comment by nathansobo** [comment]
Nicely done. Was an enjoyable diff to read. Love the per-node caching.

File: src/native/marker-index.cc (hunk 1 above)
Comment (maxbrunsfeld) [inline src/native/marker-index.cc]: :+1: Thanks for catching that. I think I started writing this thinking we could return the same variable at the end.

File: src/js/iterator.js [hunk 4]
```diff
@@ -234,6 +244,7 @@ export default class Iterator {
     while (true) {
       let comparison = compare(position, this.currentNodePosition)
 
+      this.cacheNodePosition()
```
Comment (nathansobo) [inline src/js/iterator.js]: Sorry for the late comment... is there any reason you can't call `cacheNodePosition` within `setNode`? Pretty sure that code path is exercised every time we move. I guess one reason is there are certain operations where we don't want to cache the node position when setting the node.

Comment (maxbrunsfeld) [inline src/js/iterator.js]: Yeah, I was going to put it somewhere more central originally, but thought that I'd rather avoid the overhead of doing the node caching during traversal while inserting or splicing.

I didn't try benchmarking that other approach though, so maybe we should refactor to something like that if we find that it's easier to maintain and doesn't slow things down.

Comment (nathansobo) [inline src/js/iterator.js]: Meh, this doesn't seem too tough to maintain and I think the explicit control makes sense in a data structure this critical.

**Pull request merged**